### PR TITLE
Introduced typed waves for multi data test cases

### PR DESCRIPTION
**Pull request by MichaelHuth**
In addition to the generic WAVE type multi data test cases also accept now
WAVE/T, WAVE/WAVE and WAVE/DF specific types.
The associated data generators must return waveRef waves where each element
must have the fitting type.

Added examples for all three new typed waves to DataGenSpecificWaveTypes.ipf
in internal_dev.

The MD test case data type setup is not internally buffered for each test case
found. This means that for the actual call to the test case the function
template is determined from the wave type of the element of the data generator
wave.
So for a data generator that returns a wave reference wave with text waves as
elements a test case with WAVE and with WAVE/T is valid. This applies to
reentry functions as well. See also the last example in
DataGenSpecificWaveTypes.ipf.
There is no priority applied as with a test case name only one declaration is
connected.

So the test cases entry function declaration puts constraints on the wave type of the elements from the wave reference wave from the data generator. Reentry functions are checked at runtime because they are also known through the registration. For these WAVE and the typed WAVE/xxx declaration might be valid, then both is checked and the call is done according to the found function declaration. Thus a test case entry when declared with WAVE/T constrains all elements from the data generator wave reference wave to text wave type. A following reentry function is valid with either WAVE or WAVE/T.

close https://github.com/byte-physics/igor-unit-testing-framework/issues/112

### CODE REVIEW

File: procedures/unit-testing-basics.ipf [hunk 1]
```diff
@@ -1440,9 +1484,20 @@ static Function/S CheckFunctionSignaturesTC(testCaseList, procWin)
...
+			elseif(!UTF_Utils#IsNaN(wRefSubType) && wType1 == WAVETYPE1_WREF)
+				WAVE/WAVE wWrefType = wGenerator
+				numMD = DimSize(wGenerator, UTF_ROW)
+				for(i1 = 0; i1 < numMD; i1 += 1)
+					WAVE wCheckType = wWrefType[i1]
+					if(WaveType(wCheckType, 1) != wRefSubType)
+						sprintf msg, "Test case %s expects specific wave type1 %u from the Data Generator %s. The wave type at index %u from the data generator does not fit to expected wave type.", fullTestCase, wRefSubType, dgen, i
+						UTF_PrintStatusMessage(msg)
+						Abort msg
+					endif
```
Comment (t-b) [inline procedures/unit-testing-basics.ipf:1496]: How about a sub routine here like

```
static Function HasConstantWaveTypes(wv, subType)
WAVE/WAVE wv
variable subType

Make/FREE/N=(DimSize(wv, UTF_ROWS)) matches = WaveType(wv[0], 1) == wRefSubType

FindValue/V=(0) matches
return V_Value == -1
```

and then

```
if(!HasConstantWaveTypes(wWrefType, wRefSubType))
```

File: procedures/unit-testing-basics.ipf [hunk 2]
```diff
@@ -931,6 +931,36 @@ Function TEST_CASE_PROTO_MD_WV([wv])
 	abortNow()
 End
 
+Function TEST_CASE_PROTO_MD_WVTEXT([textWave])
```
Comment (t-b) [inline procedures/unit-testing-basics.ipf:934]: Can we use `wv` as optioal parmeter name as well? This makes writing test cases which use that feature easier as I don't have to remember the specialized name.

**Review by t-b — CHANGES_REQUESTED** [review]

File: procedures/unit-testing-basics.ipf (hunk 1 above)
Comment (MichaelHuth) [reply procedures/unit-testing-basics.ipf]: ok, utility function is added and used.

File: procedures/unit-testing-basics.ipf (hunk 2 above)
Comment (MichaelHuth) [reply procedures/unit-testing-basics.ipf]: Changed optional parameters to wv as well.
Though the name in the test case/reentry functions do not matter as it is an optional parameter.

**Review by t-b — APPROVED** [review]
LGTM.

**Pull request merged**